openedx_tagging/core/tagging/rules.py

```diff
@@ -53,17 +53,20 @@ def can_change_tag(user: UserType, tag: Tag | None = None) -> bool:
 
 
 @rules.predicate
-def can_change_object_tag(user: UserType, object_tag: ObjectTag | None = None) -> bool:
+def can_change_object_tag(user: UserType, taxonomy: Taxonomy | None = None) -> bool:
     """
-    Taxonomy admins can create or modify object tags on enabled taxonomies.
+    Everyone can potentially create/edit object tags (taxonomy=None). The object permission must be checked
+    to determine if the user can create/edit a object_tag for a specific taxonomy.
```

Comment:
I think we need to emphasize this more, if I'm understanding it correctly. Something like:

```
Everyone can potentially create/edit object tags, IF they have permission
to use the taxonomy that the tag comes from, and IF they have permission
to edit the object in question.

❗This function only checks if they have permission to use the taxonomy! You
need to combine this with a separate check for their permission to edit the
object, which depends on the type of object.

The permissions for the taxonomy usage are simple: Everyone can create
or modify object tags using an enabled taxonomy. Only taxonomy admins
can create or modify object tags using a disabled taxonomies.
```

But I'm also wondering, why are admins allowed to create object tags using disabled taxonomies? Do we actually need that?

Reply:
It is a bit different. Because we are using the DjangoObjectPermission in the view, this function is called two times when we make a rest call:
1. First, for checking if the user has the METHOD permission (PUT, in this specific case). In this call, we have taxonomy=None.
2. After that, we call this function passing a taxonomy to check if the user has permission for this taxonomy

Everyone passes on 1. This is what I wanted to say with `Everyone can potentially create/edit object tags (taxonomy=None).`. I was trying to justify the `if not taxonomy: return True`

This is an implementation detail. Do you think we should let this out of the docstring?

This text covers the actual permission as of today:
```
Everyone can create or modify object tags using an enabled taxonomy. Only taxonomy admins
can create or modify object tags using a disabled taxonomy.
```

> But I'm also wondering, why are admins allowed to create object tags using disabled taxonomies? Do we actually need that?
I'm not sure. It makes sense to me to let an admin clear the tags from a disabled taxonomy, for example.

Reply:
@rpenido That makes sense, thanks. Can you please explain that in the code, either in the docstring or in comments? It wasn't super clear to me what's happening until you replied.

> This text covers the actual permission as of today:

But also we need to remind developers to check the object permissions to. We can't let just anyone apply object tags to a course, if they don't also have permission to edit that course in Studio!

Reply:
Actually, how will that work with this API? Does it have a way of checking if the user has edit permission for the object?

Reply:
> @rpenido That makes sense, thanks. Can you please explain that in the code, either in the docstring or in comments? It wasn't super clear to me what's happening until you replied.

Sure @bradenmacdonald! I'll work on this.

> Or we need to make this a base class that can be inherited and customized e.g. in edx-platform but not used directly, since it doesn't know how to check permissions. Or we need to add a pluggable permissions check.

I think this is the strategy: inherit this class in edx-platform to enhance the permission.

About the object permission, look at the discussion in the issue [here](https://github.com/openedx/modular-learning/issues/82#issuecomment-1679590226). Maybe I oversimplified reducing the permission check to only taxonomy.

In the `openedx-learning` side, as a stand-alone app, we basically apply tags to "strings" without any object knowledge.

I agree that we can have a callback to check some type of "Object" permission, but I'm not sure if we need this for the MVP.

Maybe aggregating more permissions in `edx-platform` (not on Object level,  but Taxonomy level) is enough?

CC @ChrisChV